### DISCUSSION TITLE
Interest in CIF and mmCIF format support for large biological assemblies

### DISCUSSION THREAD

**Discussion by rjboyd00** [BioPandas/biopandas#92] [Ideas]
Behavior when trying to fetch PDB structures that are too large for typical .pdb file format is as follows:

<img width="977" alt="Screen Shot 2022-03-05 at 10 12 54 PM" src="https://user-images.githubusercontent.com/6279827/156911779-5b7b90b0-a1be-4335-a02f-6fccb9b53503.png">

Word on the street is that the PDB is moving towards the mmCIF file format as the primary file format in the future and it would be great to have support for a pythonic way to interact with these file formats that plays well with large structures.

**Comment by rasbt** [comment]
Thanks for the note! Regarding the issue, I wonder if that's a server API related issue and whether it works if you manually download the file and load it from your local storage?

Regarding the bigger picture: Thanks for pointing out mmCIF. Honestly, I haven't been working with protein structures in quite some time and am not up to date there. It's definitely something worthwhile adding in the long run though. Overall, it's very exciting to hear that the PDB format is potentially to be retired soon, because I think unless you work with Fortran, it's nothing but hassle.

I.e., I can see a `PandasMmCIF` in parallel to the existing `PandasPdb` class. I am not sure if I have the time and capacity for implementing it in the short term, but I will add it to the issue tracker.

**Reply by rjboyd00** [reply]
I played around with it a bit. It looks like the `.read_pdb()` function doesn't produce the same error after I open the .cif in chimera and save it as a .pdb. Not sure what all of the NaN in my .pdb are from but it looks to me like there is also room for a really killer function to convert file types! I agree that .cif will be a huge improvement! Especially as someone who works on massive virus capsid structures. 

<img width="709" alt="Screen Shot 2022-03-08 at 4 10 49 PM" src="https://user-images.githubusercontent.com/6279827/157347338-33f4ec32-781b-40f4-a994-7cd596079385.png">

**Reply by rasbt** [reply]
Glad to hear that this one works. I think when you ran `fetch_pdb` the error was something about their website and loading the file from there (hence the `HTTP Error 404`). 

Actually, when I go to [https://files.rcsb.org/download/7rd1.pdb](https://files.rcsb.org/download/7rd1.pdb), it does seem to be unavailable. (If you replace 7rd1 by 3eiy, it works, so there is something about that particular entry). Probably there is only a .cif file, not a .pdb file as you suggested.

**Reply by a-r-j** [reply]
Yep, there is no PDB available - if you check the options under "Download Files" [here](https://www.rcsb.org/structure/7RD1) it says they're not available for especially large structures.


From the [horse's mouth](https://www.rcsb.org/docs/general-help/structures-without-legacy-pdb-format-files):

```
Several type of PDB entries are not offered in the legacy PDB format anymore:

Entries containing multiple character chain ids
Entries containing > 62 chains
Entries containing > 99999 ATOM coordinates
Entries that have complex beta sheet topology, see [more details](https://www.wwpdb.org/news/news?year=2021#60abe1a3301dc46fd700221d)
```

**Comment by pierrepo** [comment]
Hello :wave: 

regarding the mmCIF format, I was wondering if it could be possible to write a file in such format. As far as I know, biopandas can read mmCIF files but cannot write these files.

**Reply by a-r-j** [reply]
Hiya! 

Yep, that's correct. If you need to write, there is a mmcif->pdb conversion that would let you write the contents to a PDB file. However there is no mmcif format output.

I've got too much on my plate to take a look at this but would be very happy to provide support on a PR :)
